src/device/device.cpp

```diff
@@ -0,0 +1,214 @@
+//    used to endorse or promote products derived from this software without specific
+//    prior written permission.
+// 
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+// EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+// MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+// THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+// SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+// PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+// STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+// THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+
+#include "device.hpp"
+#include "device_default.hpp"
+#include "device_ledger.hpp"
+
+
+//#include "ringct/rctOps.h"
+
+#include "common/scoped_message_writer.h"
+
+namespace hw {
+
+
+    /* ======================================================================= */
+    /*  MAGIC LAMBDA SETUP                                                     */
+    /* ======================================================================= */
+
+    namespace {
+        static std::map<std::string, Device&> registry;
+        static int inited = []() -> int {
+            //tools::scoped_message_writer w  = tools::msg_writer();
+            //w << "Registering devices...";
+            hw::core::register_all();
+            #ifdef HAVE_PCSC
+            hw::ledger::register_all();
+            #endif
+            //w << "done.";
+
+            return registry.size();
+        }();
+    }
+
+    Device& register_device(const std::string device_descriptor, Device& device) {
+        //tools::scoped_message_writer w  = tools::success_msg_writer();
+        registry.insert(std::make_pair(device_descriptor,std::ref(device)));
+        return device;
+    }
+
+
+    Device& get_device(const std::string device_descriptor) {
+        auto device = registry.find(device_descriptor);
+         if (device == registry.end()) {
+            tools::fail_msg_writer()<< "Device not found "<<device_descriptor;
+            throw std::runtime_error("device not found: "+ device_descriptor);
+         }
+
+         return device->second;
+    }
+
+
+    /* ======================================================================= */
+    /*  DEVICE                                                                 */
+    /* ======================================================================= */
+
+    /* --- SETUP/TEARDOWN --- */
+    bool Device::set_name(const std::string &name) {
+        this->name = name;
+        return true;
+    }
+    const std::string Device::get_name()  const {
+        return this->name;
+    }
+    bool Device::init(void) {
+        dfns();
```

Comment:
All of these should be pure virtual, or the default implementation should be merged into this one.

Reply:
device_default has no specific implementation for get/set_name.  This is a default implementation, subclass may or may not override 

Others provide safe manner to identify a fail/missing device routing.

So I don't see what's wrong. Maube I do not understand what you say :)

Reply:
None of these have implementations. Drop the implementation entirely. In the header write `virtual void init() = 0;` instead. The compiler will then generate an error unless a subclass implements that specific function. Do that with all of these unimplemented methods, or put the default implementation into this one. There is no value in having an interface that **is** an implementation.